R/prioritize_dt.R

```diff
@@ -0,0 +1,128 @@
+#' @title Rank non-unique rows in a data.table using defined priority orders
+#'
+#' @param dt \[`data.table()`\]\cr
+#'   Data to determine rank priority for.
+#' @param rank_by_cols \[`character()`\]\cr
+#'   Apply `rank_order` priorities to each unique combination of `rank_by_cols`
+#'   in `dt`.
+#' @param rank_order \[`list()`\]\cr
+#'   Priority order to use when ranking non-unique rows. Each element of
+#'   `rank_order` corresponds to a column in `dt`. Possible values for each
+#'   column are '1' (ascending), '-1' (descending) or ordered factor levels when
+#'   the column is not a numeric. See details for more information.
+#' @param quiet \[`logical(1)`\]\cr
+#'   Whether to print out detailed messages/warnings about possible issues with
+#'   `rank_order`. Default is 'FALSE'.
+#'
+#' @return `dt` with a new 'priority' column generated using the rules specified
+#'   in `rank_order`. 'priority' equal to 1 is the highest priority
+#'
+#' @details
+#' `prioritize_dt` uses `data.table::setorderv` to order `dt` according to
+#' `rank_order`. `prioritize_dt` takes three possible values to specify the
+#' order of a column in `dt`.
+#' 1. '1', order a numeric column in ascending order (smaller values have higher
+#' priority).
+#' 2. '-1', order a numeric column in descending order (larger values have
+#' higher priority).
+#' 3. `factor` levels, to order a categorical column in a custom order with the
+#' first level having highest priority. When not all present values of the
+#' column are defined in the levels, the priority will be NA and a warning
+#' printed if `quiet = FALSE`.
+#'
+#' @examples
+#' # preliminary data with only total population
+#' dt_total <- data.table::CJ(
+#'   location = "USA", year = 2000, age_start = 0, age_end = Inf,
+#'   method = c("de facto", "de jure"),
+#'   status = c("preliminary")
+#' )
+#' # final data in 10 year age groups
+#' dt_10_yr_groups <- data.table::CJ(
+#'   location = "USA", year = 2000, age_start = seq(0, 80, 10),
+#'   method = c("de facto", "de jure"),
+#'   status = c("final")
+#' )
+#' dt_10_yr_groups[, age_end := age_start + 10]
+#' dt_10_yr_groups[age_start == 80, age_end := Inf]
+#'
+#' input_dt <- rbind(dt_total, dt_10_yr_groups)
+#' input_dt[, n_age_groups := .N, by = setdiff(names(input_dt), c("age_start", "age_end"))]
+#'
+#' output_dt <- prioritize_dt(
+#'   dt = input_dt,
+#'   rank_by_cols = c("location", "year"),
+#'   rank_order = list(
+#'     method = c("de facto", "de jure"), # prioritize 'de facto' sources highest
```

Comment:
> How are ties managed? So if you had one source that's de facto w/ fewer age groups and one source that's de jure with more age groups, for example. I think this could be operationalized by saying the rules that are more important go first in the `rank_order` list.

Yup the order of `rank_order` matters, tried to clarify this in the detailed documentation.

Reply:
> Also -- what happens if you have two different sources that are de facto and have the maximum number of age groups? What does it look like to use this function to just say "in general we prefer X source over Y source"? Maybe make `source` the last tie breaker rule in `rank_order` with factored variables for the most common sources.

Good question, right now they would both be assigned 1st priority which we wouldn't want. I think in that case we'd want to error out and flag that an additional identifying variable is required (like `source` in the example you gave).

Reply:
Okay so I added an additional check to make sure once the ranking is applied `dt` is uniquely identified by a given set of id cols and the `priority` column. Which with your example would flag that `source` needs to be added to `rank_order`.